### PR TITLE
DeepSeek V4 Bridge

### PR DESCRIPTION
**Pull request by weijiac0619**
## Summary
- Adds DeepSeek V4 bridge support using native Hugging Face DeepSeek V4 configs.
- Adds FP8/MXFP4 checkpoint import/export mappings, docs/examples, and focused coverage.
- Pins the MCore dev commit with DeepSeek V4 Part 3 support and refreshes `uv.lock`.

## Validation
- Full HSG EP8 import and export roundtrip completed.
- `pre-commit run --all-files` passed.
- `uv lock --check` and locked `uv sync` dry-runs passed.

### CODE REVIEW

File: src/megatron/bridge/models/conversion/model_bridge.py [hunk 1]
```diff
@@ -920,6 +920,9 @@ def load_weights_hf_to_megatron(
 
         _hf_import_cache: Dict[str, torch.Tensor] = {}
         for task in self._with_progress_tracking(hf_to_megatron_tasks, description):
+            # task is None when the param has no mapping or its HF key is absent from the checkpoint
+            if task is None:
+                continue
```
Comment (cuichenx) [inline src/megatron/bridge/models/conversion/model_bridge.py]: this shouldn't be needed -- i think this indicates the mapping is wrong?

File: src/megatron/bridge/models/glm_moe_dsa/glm5_bridge.py [hunk 2]
```diff
@@ -15,7 +15,12 @@
 import logging
 
 from megatron.core.models.gpt.gpt_model import GPTModel
-from transformers import GlmMoeDsaForCausalLM
+
+
+try:
+    from transformers import GlmMoeDsaForCausalLM
+except ImportError:
+    GlmMoeDsaForCausalLM = None
```
Comment (cuichenx) [inline src/megatron/bridge/models/glm_moe_dsa/glm5_bridge.py]: upgrading to latest transformers should resolve this issue

File: src/megatron/bridge/models/deepseek/deepseek_v4_bridge.py [hunk 3]
```diff
@@ -0,0 +1,905 @@
...
+        super().__init__(**kwargs)
+
+
+# Register at import time so AutoConfig.from_pretrained works before any bridge
+# method is called.
+try:
+    AutoConfig.register("deepseek_v4", DeepseekV4Config)
+except ValueError as e:
+    if "already registered" not in str(e):
+        raise
```
Comment (cuichenx) [inline src/megatron/bridge/models/deepseek/deepseek_v4_bridge.py]: don't think this block is needed? we can use `DeepseekV4ForCausalLM`

File: src/megatron/bridge/models/deepseek/deepseek_v4_bridge.py [hunk 4]
```diff
@@ -0,0 +1,905 @@
...
+
+        # ---- Hyper-Connections (mHC) ----
+        provider.enable_hyper_connections = True
+        provider.num_residual_streams = hf_config.hc_mult  # 4
+        provider.mhc_sinkhorn_iterations = hf_config.hc_sinkhorn_iters  # 20
+
+        # ---- MoE ----
+        provider.gated_linear_unit = True
+        provider.moe_grouped_gemm = True
+        provider.moe_router_pre_softmax = False  # V4 uses post-topk normalisation
```
Comment (cuichenx) [inline src/megatron/bridge/models/deepseek/deepseek_v4_bridge.py:415]: Can we verify this? v3 used pre softmax

File: examples/models/deepseek_v4/README.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+End-to-end conversion and inference scripts for the DeepSeek V4 family on Megatron Bridge.
+
+The bridge supports four published variants out of the same code path. The on-disk quantisation differs between post-trained (Flash, Pro) and pretrained-only (Flash-Base, Pro-Base) models — see [`docs/models/llm/deepseek-v4.md`](../../../docs/models/llm/deepseek-v4.md) for the per-variant scheme.
+
+| Variant | HF path | Quant scheme | TP | EP | Min GPUs | Parity status |
+|---------|---------|--------------|---:|---:|---------:|---------------|
+| DeepSeek-V4-Flash | `deepseek-ai/DeepSeek-V4-Flash` | FP8 attn + MXFP4 experts | 1 | 4 | 4 × B200 192 GB | Verified, last-real-token logit cosine ~0.97-0.99 vs official inference |
+| DeepSeek-V4-Flash-Base | `deepseek-ai/DeepSeek-V4-Flash-Base` | uniform FP8 (F32 scales) | 1 | 4 | 4 × B200 192 GB | Verified, last-real-token logit cosine 0.9866-0.9930, mean 0.9907 vs official inference |
+| DeepSeek-V4-Pro | `deepseek-ai/DeepSeek-V4-Pro` | FP8 attn + MXFP4 experts | 1 | 16 | 16 × B200 192 GB | Algorithmic dequant validated; end-to-end unmeasured |
```
Comment (cuichenx) [inline examples/models/deepseek_v4/README.md]: could you update with your latest numbers

File: examples/models/deepseek_v4/README.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+## Parallelism Configurations
+
+DSv4 currently requires **TP=1** because MLA tensor parallelism is not supported alongside the DSv4 hybrid attention path. Scale via expert and pipeline parallelism instead.
+
+| Model | TP | PP | EP | Min GPUs | GPU Type | Use Case |
+|-------|---:|---:|---:|---------:|----------|----------|
+| DeepSeek-V4-Flash | 1 | 1 | 4 | 4 | B200 192 GB | Smoke / single-node inference |
+| DeepSeek-V4-Flash-Base | 1 | 1 | 4 | 4 | B200 192 GB | Smoke / single-node inference |
+| DeepSeek-V4-Pro | 1 | 1+ | 16 | 16 | B200 192 GB | Multi-node inference or conversion |
+| DeepSeek-V4-Pro-Base | 1 | 1+ | 16 | 16 | B200 192 GB | Multi-node inference or conversion |
```
Comment (cuichenx) [inline examples/models/deepseek_v4/README.md]: check if these are still accurate?

File: examples/models/deepseek_v4/README.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+```
+
+Use `./scripts/switch_mcore.sh main` and `uv sync --locked` to return to the pinned main-branch submodule.
+
+| Variant | HF path | Quant scheme | TP | EP | Validation |
+|---------|---------|--------------|---:|---:|------------|
+| DeepSeek-V4-Flash | `deepseek-ai/DeepSeek-V4-Flash` | FP8 attn + MXFP4 experts | 1 | 4 | Verified on GB200, last-real-token logit cosine ~0.97-0.99 vs official inference |
+| DeepSeek-V4-Flash-Base | `deepseek-ai/DeepSeek-V4-Flash-Base` | uniform FP8 (F32 scales) | 1 | 4 | Verified on GB200, last-real-token logit cosine 0.9866-0.9930, mean 0.9907 vs official inference |
+| DeepSeek-V4-Pro | `deepseek-ai/DeepSeek-V4-Pro` | FP8 attn + MXFP4 experts | 1 | 16 | Algorithmic dequant validated; end-to-end unmeasured |
+| DeepSeek-V4-Pro-Base | `deepseek-ai/DeepSeek-V4-Pro-Base` | uniform FP8 (F32 scales) | 1 | 16 | Algorithmic dequant validated; end-to-end unmeasured |
```
Comment (cuichenx) [inline examples/models/deepseek_v4/README.md]: could you update with the latest cosine similarity

File: examples/models/deepseek_v4/README.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+## Parallelism Configurations
+
+DSv4 currently requires **TP=1** because MLA tensor parallelism is not supported alongside the DSv4 hybrid attention path. Scale via expert and pipeline parallelism instead.
+
+| Model | TP | PP | EP | Verified Layout | Use Case |
+|-------|---:|---:|---:|-----------------|----------|
+| DeepSeek-V4-Flash | 1 | 1 | 4 | 4 GPUs on GB200 | Smoke / single-node inference |
+| DeepSeek-V4-Flash-Base | 1 | 1 | 4 | 4 GPUs on GB200 | Smoke / single-node inference |
+| DeepSeek-V4-Pro | 1 | 1+ | 16 | Not end-to-end verified | Multi-node inference or conversion |
+| DeepSeek-V4-Pro-Base | 1 | 1+ | 16 | Not end-to-end verified | Multi-node inference or conversion |
```
Comment (cuichenx) [inline examples/models/deepseek_v4/README.md]: is this still accurate?

**Comment by cuichenx** [comment]
testing ci with dev commit

File: examples/models/deepseek_v4/conversion.sh [hunk 9]
```diff
@@ -46,9 +46,11 @@ EXPORT_DIR="${WORKSPACE}/models/${MODEL_VARIANT}-hf-export"
 ITER=iter_0000000
 
 # 1) Import HF -> Megatron (FP8 / MXFP4 dequantised to bfloat16 in-flight)
-uv run python examples/conversion/convert_checkpoints.py import \
+uv run python -m torch.distributed.run --nproc_per_node=$((TP * PP * EP)) \
```
Comment (cuichenx) [inline examples/models/deepseek_v4/conversion.sh]: don't think this should be TP * PP * EP (maybe it should be PP * EP?

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3562/commits/6793a6250a962b816614572513ecf019d5c7dba1

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3562/commits/f09cd63c552f2c268831290fd09e5c5460af17c0

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3562/commits/1986ee8a5369d0d061a45299133b75f85e5b8a10

**Comment by cuichenx** [comment]
main passed (dsv4 tests skipped): https://github.com/NVIDIA-NeMo/Megatron-Bridge/actions/runs/26012504739/job/76468557985?pr=3562

dev ci running now

**Comment by cuichenx** [comment]
dev ci passed with only one known failure about deepseek v3 training. Current PR does not affect training. 
DSV4 tests were not skipped. 
https://github.com/NVIDIA-NeMo/Megatron-Bridge/actions/runs/26068735913/job/76660085078

File: src/megatron/bridge/models/deepseek/deepseek_v4_bridge.py [hunk 10]
```diff
@@ -0,0 +1,952 @@
...
+  - embed.weight            (not model.embed_tokens.weight)
+  - head.weight             (not lm_head.weight)
+  - norm.weight             (not model.norm.weight)
+  - layers.N.attn_norm.weight / layers.N.ffn_norm.weight
+  - layers.N.attn.wq_a / wq_b / wkv / wo_a / wo_b …
+  - layers.N.ffn.gate / experts / shared_experts …
+  - layers.N.hc_attn_fn / hc_attn_base / hc_attn_scale  (Hyper-Connections)
+  - layers.N.hc_ffn_fn  / hc_ffn_base  / hc_ffn_scale
+  - hc_head_fn / hc_head_base / hc_head_scale            (global HC head, learned output contraction)
+  - mtp.N.*                                               (MTP layers)
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/deepseek/deepseek_v4_bridge.py:34]: is mtp working?

File: examples/models/deepseek_v4/README.md [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+| Model | TP | PP | EP | GPUs | GPU | Verified |
+|-------|---:|---:|---:|-----:|-----|----------|
+| DeepSeek-V4-Flash | 1 | 1 | 4 | 4 | GB200 192GB | Import, export, inference |
+| DeepSeek-V4-Flash | 1 | 1 | 8 | 8 | H100 80GB | Import, export, inference |
+| DeepSeek-V4-Flash-Base | 1 | 1 | 4 | 4 | GB200 192GB | Import, export, inference |
+| DeepSeek-V4-Pro | 1 | 4 | 8 | 32 | GB200 192GB | Import, export, inference |
+| DeepSeek-V4-Pro | 1 | 16 | 8 | 128 | H100 80GB | Import, export, inference |
+
+## Known Limitations
```
Comment (yaoyu-33) [inline examples/models/deepseek_v4/README.md:51]: if there are on going mcore PR not merged, can we attatch link here? we can clean up later.

File: src/megatron/bridge/models/deepseek/deepseek_v4_bridge.py (hunk 10 above)
Comment (cuichenx) [reply src/megatron/bridge/models/deepseek/deepseek_v4_bridge.py:34]: yep, both conversion and training

**Review by cuichenx — DISMISSED** [review]

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3562/commits/a9b3edd272b3df8ce6eba972f368476ee13ada51

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3562/commits/fe9131e9903886b5e3343597b83af3b072b6d5a4

**Comment by cuichenx** [comment]
main ci passed again besides coverage. merging
